apm-sniffer/bootstrap-plugins/jdk-threadpool-plugin/src/main/java/org/apache/skywalking/apm/plugin/wrapper/SwRunnableWrapper.java

```diff
@@ -0,0 +1,52 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+package org.apache.skywalking.apm.plugin.wrapper;
+
+import org.apache.skywalking.apm.agent.core.context.ContextManager;
+import org.apache.skywalking.apm.agent.core.context.ContextSnapshot;
+import org.apache.skywalking.apm.agent.core.context.trace.AbstractSpan;
+import org.apache.skywalking.apm.network.trace.component.ComponentsDefine;
+
+public class SwRunnableWrapper implements Runnable {
+
+    private Runnable runnable;
+
+    private ContextSnapshot contextSnapshot;
+
+    public SwRunnableWrapper(Runnable runnable, ContextSnapshot contextSnapshot) {
+        this.runnable = runnable;
+        this.contextSnapshot = contextSnapshot;
+    }
+
+    @Override
+    public void run() {
+        AbstractSpan span = ContextManager.createLocalSpan(getOperationName());
```

Comment:
Are you considering `spin-tracing`? The agent is going to trace its own internal methods? And meanwhile, this method could print logs. You will face endless executions.

Reply:
> 

I  have considered this case. the instructed point are  just`execute` and `submit` methods  in `ThreadPoolExecutor` and its subclasses.  `FileWriter` and `JVMService` use `scheduleAtFixedRate` method.   

If the `execute` method in `ThreadPoolExecutor` subclass calls its super class `execute` method, this would result in `Runnable` will be wrapped twice, but it won't form a `spin-tracing`.

Reply:
But `scheduleAtFixedRate` is also could be used by users, which would be missed by this plugin.

Reply:
I think the better way is to identify the caller and exclude the agent package. WDYT?
How to identify could be discussed later.

Reply:
> 

To be honest, My focus is `ThreadPoolExecutor`, not all kinds of 'TheadPool`s

Reply:
> I think the better way is to identify the caller and exclude the agent package. WDYT? How to identify could be discussed later.

Ok, I will think another better way. Keep in touch!

Reply:
2 general ways.
1. Ask SkyWalking core to use a specific interface(extends Runnable/Callable) only. Then the plugin is easy to resolve the caller.
2. Get the stack and try to find `org.apache.skywalking`. 

The (2) is easier but it costs performance due to getting stack elements.

Reply:
> 

Yes, `ScheduledThreadPool` is inherited from `ThreadPoolExecutor`, also has `execute` and `submit` methods. so I take all the subclasses of `ThreadPoolExecutor` into account, but only `execute` and `submit` methods. I'll reconsider it all

Reply:
> Yes, `ScheduledThreadPool` is inherited from `ThreadPoolExecutor`, also has `execute` and `submit` methods. so I take all the subclasses of `ThreadPoolExecutor` into account, but only `execute` and `submit` methods. I'll reconsider it all

Yes, ScheduledThreadPool is inherited from ThreadPoolExecutor, also has execute and submit methods. so I take all the subclasses of ThreadPoolExecutor into account, but only execute and submit methods. Do you think we just take `scheduleAtFixedRate` and `scheduleWithFixedDelay` into account in addition to `execute` and `submit`, or cover all kinds of threadpools?

Reply:
Once this is `ThreadPoolExecutor` plugin, it should cover all known cases, which is why we should detect caller one way or another.

Reply:
And, have you known there are meter plugins for thread pool, such as `UndertowWorkerThreadPoolConstructorIntercept`. I think if this plugin covers the thread pool case too, it should report the thread pool metrics too.

Reply:
<img width="449" alt="image" src="https://user-images.githubusercontent.com/5441976/162208090-da5268fa-95d2-4970-bef7-9c6b90165324.png">

This is available on v9 `General Service/Instance/JVM` panel.

Reply:
> And, have you known there are meter plugins for thread pool, such as `UndertowWorkerThreadPoolConstructorIntercept`. I think if this plugin covers the thread pool case too, it should report the thread pool metrics too.

I've seen the issue before. I seen it again just now. it also instructs `ThreadPoolExecutor` with  the match rule that package name is started with "org.xnio". If my `ThreadPoolExecutor` plugin covers  this, the `Undertow ThreadPool` plugin is redundant I think. I will take it into account. 

Please take a look at [the reply](#discussion_r845092153) and give me a reply. Thank you!

Reply:
> My focus is ThreadPoolExecutor, not all kinds of TheadPools.

This makes perfect sense to me. We don't need to cover all threads.

> Do you think we just take scheduleAtFixedRate and scheduleWithFixedDelay into account in addition to execute and submit, or cover all kinds of threadpools?

Not all thread pool, just these two popular scheduling ways.

Reply:
@wu-sheng  `ScheduledThreadPoolExecutor.scheduleAtFixedRate` has been supported. but  the trace is getting longer and longer over the time. I think it is a bad sense. WDYT?
![image](https://user-images.githubusercontent.com/29882506/162677216-12b2c6d2-0125-42a6-bb0b-4e4b708de138.png)

Besides, `ScheduledThreadPoolExecutor.scheduleAtFixedRate` in `JVMService` doesn't show in the trace ui. I don't identify the caller.  I don't know the reason.

Reply:
OK, I think we should not support this for now. It makes the trace strange.

Reply:
> OK, I think we should not support this for now. It makes the trace strange.

Agree.

Reply:
> > And, have you known there are meter plugins for thread pool, such as `UndertowWorkerThreadPoolConstructorIntercept`. I think if this plugin covers the thread pool case too, it should report the thread pool metrics too.
> 
> I've seen the issue before. I seen it again just now. it also instructs `ThreadPoolExecutor` with the match rule that package name is started with "org.xnio". If my `ThreadPoolExecutor` plugin covers this, the `Undertow ThreadPool` plugin is redundant I think. I will take it into account.
> 
> Please take a look at [reply1](#discussion_r845039466), [reply2 ](#discussion_r845092153). I want to know your point of view. Thank you!

@wu-sheng  I am adding `ThreadPoolExecutor` metrics. But How to distinguish different thread pools？That is, how to determine the thread pool name. so far, I haven't thought of a meaningful name yet. Do you have any good suggestions?

Reply:
Thread pool supports setting a custom name by users. AFAIK.

Reply:
If it is not set, I think we could use pool's object ID as the name? It is meaningless, yes, but it is better than nothing.
After all, the thread pool metric is at instance level only.

Reply:
> Thread pool supports setting a custom name by users. AFAIK.

User can custom thread's name, not threadpool's name. They are not equivalent. I will use object ID or other name that seems more meaningful.

Reply:
@wu-sheng Usually, an application includes many thread pools. If we collect all thread pool meters, but users don't pay attention to these indicators, then these collected meters are garbage data. I think we can let users configure thread pools they care about in `agent.config`. the rough format is  `class1:name1,name2;class2:name1,name2`. Then we can match it  and name `ThreadFactory` in `ThreadPoolExecutor` plugin through `thread stack`. Although it consumes a little performance when the application starts, I think it's worth it. WDYT?

Reply:
Got it, make sense. Yes, maybe we have to use the class name as the thread pool name. 

I think we could make trace supported in this PR and do meter collecting in another?

Reply:
One more scenario to discussion, 

The ThreadPoolExecutor could be used to run an endless loop task, like a timer. And `run` method would never end. 
As you mentioned using the class name to filter in meter collection, should we add this kind of mechanism in the tracing part too? (Have to back about caller detection, as user could use APIs rather than inherit from this)

Reply:
> Got it, make sense. Yes, maybe we have to use the class name as the thread pool name.
> 
> I think we could make trace supported in this PR and do meter collecting in another?

ok, agree!

Reply:
> One more scenario to discussion,
> 
> The ThreadPoolExecutor could be used to run an endless loop task, like a timer. And `run` method would never end. As you mentioned using the class name to filter in meter collection, should we add this kind of mechanism in the tracing part too? (Have to back about caller detection, as user could use APIs rather than inherit from this)

Do you mean we do not trace some `Runnable` or `Callable` task ?

Reply:
No need, please address others. This is good for me now.